### PR TITLE
Added OpenCL CPU and GPU to Jenkins Pipeline

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py src/test/unit`
- [ ] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This PR will add OpenCL testing capabilities to our Stan Jenkins Pipeline for CPU and GPU as requested in https://github.com/stan-dev/stan/pull/3048

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Nicusor Serban, Toptal.

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -192,6 +193,9 @@ pipeline {
                     ].join(" ")
 
                     skipRemainingStages = utils.verifyChanges(paths)
+
+                    def openCLPaths = ['lib/stan_math/stan/math/opencl', 'lib/stan_math/test/unit/math/opencl'].join(" ")
+                    skipOpenCL = utils.verifyChanges(openCLPaths)
```
Comment (rok-cesnovar) [inline Jenkinsfile:198]: This should just check if there are  changes in `src/stan/model/indexing`, we dont really care about changes in math here.

File: Jenkinsfile [hunk 2]
```diff
@@ -392,6 +396,81 @@ pipeline {
...
+                                deleteDir()
+                                unstash 'StanSetup'
+                                sh "echo CXX=${env.CXX} -Werror > make/local"
+                                sh "echo STAN_OPENCL=true>> make/local"
+                                sh "echo OPENCL_PLATFORM_ID=${env.OPENCL_PLATFORM_ID_CPU}>> make/local"
+                                sh "echo OPENCL_DEVICE_ID=${env.OPENCL_DEVICE_ID_CPU}>> make/local"
+                                // skips tests that require specific support in OpenCL
+                                sh 'echo "ifdef NO_CPU_OPENCL_INT64_BASE_ATOMIC" >> make/local'
+                                sh 'echo "CXXFLAGS += -DSTAN_TEST_SKIP_REQUIRING_OPENCL_INT64_BASE_ATOMIC" >> make/local'
+                                sh 'echo "endif" >> make/local'
```
Comment (rok-cesnovar) [inline Jenkinsfile:425]: The above 3 lines can be removed (together with the comment).

File: Jenkinsfile [hunk 3]
```diff
@@ -392,6 +396,81 @@ pipeline {
...
+                                runTests("lib/stan_math/test/unit/math/opencl", false)
```
Comment (rok-cesnovar) [inline Jenkinsfile:426]: The test we will need to run are just

```cpp
runTests("src/test/unit")
```
We actually only need to run assign_cl and rvalue_cl, but these take just a few minutes so its fine.

File: Jenkinsfile [hunk 4]
```diff
@@ -392,6 +396,81 @@ pipeline {
...
+                                deleteDirWin()
+                                unstash 'StanSetup'
+                                bat "echo CXX=${env.CXX} -Werror > make/local"
+                                bat "echo STAN_OPENCL=true >> make/local"
+                                bat "echo OPENCL_PLATFORM_ID=${env.OPENCL_PLATFORM_ID_CPU} >> make/local"
+                                bat "echo OPENCL_DEVICE_ID=${env.OPENCL_DEVICE_ID_CPU} >> make/local"
+                                bat 'echo LDFLAGS_OPENCL= -L"C:\\Program Files (x86)\\IntelSWTools\\system_studio_2020\\OpenCL\\sdk\\lib\\x64" -lOpenCL >> make/local'
+                                bat "mingw32-make.exe -f make/standalone math-libs"
+                                runTestsWin("lib/stan_math/test/unit/math/opencl", false, false)
+                                runTestsWin("lib/stan_math/test/unit/multiple_translation_units_test.cpp", false, false)
```
Comment (rok-cesnovar) [inline Jenkinsfile:438]: Replace the two with `runTestsWin("src/test/unit")`

**Review by rok-cesnovar — COMMENTED** [review]
Nic, this looks great. Just a few pointers of what needs to be tweaked compared to what we run in Math.

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar thanks for the quick reply! All comments solved.

**Comment by serban-nicusor-toptal** [comment]
Updated base branch to `opencl_indexing` so we can target t4c1's PR.

**Comment by serban-nicusor-toptal** [comment]
It seems to have failed in `./runTests.py -j8 src/test/unit --make-only` see [logs](https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-3050/4/pipeline/)

**Review by rok-cesnovar — APPROVED** [review]
I am approving the Jenkinsfile changes. 

@SteveBronder I added the stancflags parts to the makefile. Just copied from cmdstan and its easy but please double check.

Not sure why the overrides popped up here.

**Review by SteveBronder — APPROVED** [review]
Thanks looks great!

**Comment by SteveBronder** [comment]
@rok-cesnovar is this a problem in the compiler?

https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-3050/13/pipeline#step-145-log-179

**Comment by rok-cesnovar** [comment]
It would appear that way, but I have no issue compiling this model https://github.com/stan-dev/stan/blob/develop/src/test/test-models/good/mcmc/hmc/hamiltonians/funnel.stan locally with STAN_OPENCL on. As well as the test in question also passes. Will look into this a bit more. Might be clang++-6.0 issue?

**Comment by rok-cesnovar** [comment]
Some tests seem to also test with fvar. Added guards to not run those with OpenCL and we whould now finally be good to go here. Will let you approve.

**Pull request merged**